lenskit-core/src/main/java/org/lenskit/data/entities/AbstractEntity.java

```diff
@@ -0,0 +1,218 @@
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.data.entities;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Iterators;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.*;
+
+/**
+ * Base class to make it easier to implement entities.
+ */
+public abstract class AbstractEntity implements Entity {
+    @Override
+    public Collection<AttributeValue<?>> getAttributeValues() {
+        return new ValueCollection();
+    }
+
+    @Override
+    public Map<String, Object> asMap() {
+        return new MapView();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This implementation delegates to {@link #maybeGet(Attribute)}.
+     */
+    @Nonnull
+    @Override
+    public <T> T get(Attribute<T> attribute) {
+        T val = maybeGet(attribute);
+        if (val == null) {
+            throw new NoSuchAttributeException(attribute.toString());
+        } else {
+            return val;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This implementation delegates to {@link #maybeGet(String)}.
+     */
+    @Nonnull
+    @Override
+    public Object get(String attr) {
+        Object val = maybeGet(attr);
+        if (val == null) {
+            throw new NoSuchAttributeException(attr);
+        } else {
+            return val;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This implementation delegates to {@link #get(Attribute)}.
+     */
+    @Override
+    public long getLong(Attribute<Long> attribute) {
+        return get(attribute);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This implementation delegates to {@link #get(Attribute)}.
+     */
+    @Override
+    public double getDouble(Attribute<Double> attr) {
+        return get(attr);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        } else if (obj instanceof Entity) {
+            Entity oe = (Entity) obj;
+            return getType().equals(oe.getType()) && getId() == oe.getId() && asMap().equals(oe.asMap());
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        HashCodeBuilder hcb = new HashCodeBuilder();
+        hcb.append(getType())
+           .append(getId())
+           .append(asMap());
+        return hcb.toHashCode();
+    }
+
+    @Override
+    public String toString() {
+        ToStringBuilder tsb = new ToStringBuilder(this, ToStringStyle.SHORT_PREFIX_STYLE);
+        tsb.append("type", getType())
+           .append("id", getId());
+        for (AttributeValue<?> av: getAttributeValues()) {
```

Comment:
given that I remember seeing a definition of identifier as an attribute, how likely is is that ID shows up twice?